### PR TITLE
[SPARK-22063][R] Fixes lint check failures in R by latest commit sha1 ID of lint-r

### PR DESCRIPTION
**Pull request by HyukjinKwon**
## What changes were proposed in this pull request?

Currently, we set lintr to jimhester/lintr@a769c0b (see [this](https://github.com/apache/spark/commit/7d1175011c976756efcd4e4e4f70a8fd6f287026) and [SPARK-14074](https://issues.apache.org/jira/browse/SPARK-14074)).

I first tested and checked lintr-1.0.1 but it looks many important fixes are missing (for example, checking 100 length). So, I instead tried the latest commit, https://github.com/jimhester/lintr/commit/5431140ffea65071f1327625d4a8de9688fa7e72, in my local and fixed the check failures.

It looks it has fixed many bugs and now finds many instances that I have observed and thought should be caught time to time, here I filed [the results](https://gist.github.com/HyukjinKwon/4f59ddcc7b6487a02da81800baca533c).

The downside looks it now takes about 7ish mins, (it was 2ish mins before) in my local.

## How was this patch tested?

Manually, `./dev/lint-r` after manually updating the lintr package.

### CODE REVIEW

File: R/pkg/.lintr
Comment (HyukjinKwon) [inline R/pkg/.lintr]: `object_name_linter = NULL` looks required. Otherwise, it complains about inconsistent naming styles.

File: R/pkg/R/DataFrame.R
Comment (HyukjinKwon) [inline R/pkg/R/DataFrame.R]: Looks this hits 30 length identifier limit but we can't change as exposed in the doc.

File: R/pkg/R/column.R
Comment (HyukjinKwon) [inline R/pkg/R/column.R]: I checked this link is not broken for sure.

File: R/pkg/R/context.R
Comment (HyukjinKwon) [inline R/pkg/R/context.R]: Ditto: 30 length identifier limit but exposed in the doc

File: R/pkg/R/mllib_tree.R
Comment (HyukjinKwon) [inline R/pkg/R/mllib_tree.R]: links were checked

File: R/pkg/R/mllib_tree.R
Comment (HyukjinKwon) [inline R/pkg/R/mllib_tree.R]: links were checked

File: R/pkg/R/mllib_tree.R
Comment (HyukjinKwon) [inline R/pkg/R/mllib_tree.R]: links were checked

File: dev/lint-r.R
Comment (HyukjinKwon) [inline dev/lint-r.R]: Here, I didn't yet change this. Up to my knowledge, this doesn't actually try to install this though.. just in case.

**Comment by HyukjinKwon** [comment]
@felixcheung and @shivaram,

Should we upgrade this to the latest commit? it increases the time by 5ish mins.

If there is any worry about upgrading, I am willing to leave the script as was. The other fixes look reasonable and I have wondered why these are not being caught. Does this make sense to both of you too?

**Comment by shivaram** [comment]
@HyukjinKwon Thanks for looking at this. The 5 min addition seems unfortunate though -- does that also happen with lintr-1.0.1 ? I wonder if we are seeing some specific performance slowdown because of not using a release (it might be good to follow up with lint-r authors if that is the case ?)

File: R/pkg/R/DataFrame.R
Comment (felixcheung) [inline R/pkg/R/DataFrame.R]: what's the problem here? name too long I think?

File: R/pkg/R/DataFrame.R
Comment (felixcheung) [reply R/pkg/R/DataFrame.R]: wait... this isn't public though; it's not in https://github.com/apache/spark/blob/master/R/pkg/NAMESPACE
looks like this shouldn't be in the doc (`@noRd` or `#` instead of `#'`)

File: R/pkg/R/context.R
Comment (felixcheung) [reply R/pkg/R/context.R]: this one is an API...

File: dev/lint-r.R
Comment (felixcheung) [reply dev/lint-r.R]: because of `if ("lintr" %in% row.names(installed.packages())  == FALSE) {`
?
it might be the case the Jenkins boxes already have a version installed and this won't update it

**Review by felixcheung — COMMENTED** [review]
I wonder if we should use the git tag of the 1.0.1 release
https://github.com/jimhester/lintr/releases

**Comment by felixcheung** [comment]
I think this is great to have, thanks for solving the mystery.
- 5 min: this is mildly concerning, is it possible this is caused by new checks in lintr? perhaps we could exclude them or something?
- lintr version: I see what you say about 1.0.1. According this commit list https://github.com/jimhester/lintr/commits/master the tag for 1.0.1 vs the latest (the git tag you are using) are different only by some insignificant changes?

File: dev/lint-r.R
Comment (HyukjinKwon) [reply dev/lint-r.R]: Yes .. I think I should add few more changes here if we are going to update this package.

Comment (viirya) [reply dev/lint-r.R]: It looks like the commented is a old code. Maybe a comment to explain it.

Comment (HyukjinKwon) [reply dev/lint-r.R]: Oh, just to be clear, I didn't change this yet but just commented this for now ... Actually, `devtools::install_github("jimhester/lintr@5431140")` is new one but I didn't change this yet just in case.

Will add some comments around this when I add few more changes and clean up soon in any event.

**Comment by HyukjinKwon** [comment]
To @shivaram:

> does that also happen with lintr-1.0.1 ? I wonder if we are seeing some specific performance slowdown because of not using a release (it might be good to follow up with lint-r authors if that is the case ?)

I just ran it with lintr-1.0.1. This took 2ish mins in my local, which is almost identical with the current, jimhester/lintr@a769c0b. Will open up an issue there and ask this in their repo tomorrow _(pending ..)_.

</br>
</br>

To @felixcheung:

> 5 min: this is mildly concerning, is it possible this is caused by new checks in lintr? perhaps we could exclude them or something?

I tried to disable all the linters documented in https://github.com/jimhester/lintr/blob/master/README.md#available-linters:

```diff
diff --git a/R/pkg/.lintr b/R/pkg/.lintr
index ae50b28ec61..ca1efcfacc7 100644
--- a/R/pkg/.lintr
+++ b/R/pkg/.lintr
@@ -1,2 +1,2 @@
-linters: with_defaults(line_length_linter(100), multiple_dots_linter = NULL, camel_case_linter = NULL, open_curly_linter(allow_single_line = TRUE), closed_curly_linter(allow_single_line = TRUE))
+linters: with_defaults(commas_linter = NULL, object_usage_linter = NULL, absolute_path_linter = NULL, nonportable_path_linter = NULL, pipe_continuation_linter = NULL, assignment_linter = NULL, closed_curly_linter = NULL, extraction_operator_linter = NULL, implicit_integer_linter = NULL, infix_spaces_linter = NULL, line_length_linter = NULL, no_tab_linter = NULL, object_length_linter = NULL, object_name_linter = NULL, open_curly_linter = NULL, semicolon_terminator_linter = NULL, single_quotes_linter = NULL, spaces_inside_linter = NULL, spaces_left_parentheses_linter = NULL, todo_comment_linter = NULL, trailing_blank_lines_linter = NULL, trailing_whitespace_linter = NULL, T_and_F_symbol_linter = NULL, undesirable_function_linter = NULL, undesirable_operator_linter = NULL, unneeded_concatenation_linter = NULL)
 exclusions: list("inst/profile/general.R" = 1, "inst/profile/shell.R")
```

and this took 1.5 mins. Looks we could identify some linters that we might not need. Let me check this one tomorrow _(pending ..)_

> lintr version: I see what you say about 1.0.1. According this commit list https://github.com/jimhester/lintr/commits/master the tag for 1.0.1 vs the latest (the git tag you are using) are different only by some insignificant changes?

Diff looks 122 commits - https://github.com/jimhester/lintr/compare/v1.0.1...master . Will take a look though the commit logs _(pending ..)_

**Comment by felixcheung** [comment]
isn't it the other way around?
https://github.com/jimhester/lintr/compare/v1.0.1?expand=1

**Comment by felixcheung** [comment]
btw, could you check if haven't already, if `nolint` around the `http` link, roxygen is going to handle that correctly?

**Comment by HyukjinKwon** [comment]
I initially did this, for example,

```
\href{https://spark.apache.org/docs/latest/sparkr.html#data-type-mapping-between-
r-and-spark}{Spark Data Types} for available data types.
```

this passes the lint and doc is find but cran check is failed. I actually tried to find out the way but ended up with `nolint`. ... will try to read the doc one more and out few more cases locally.

**Comment by HyukjinKwon** [comment]
Doh, you mean the current status. Yes, I checked.

**Comment by HyukjinKwon** [comment]
Will update the comment tomorrow.

**Comment by HyukjinKwon** [comment]
> anyway I mean if we use the git tag 9951084 it should match v1.0.1 release exactly.
the one you listed jimhester/lintr@5431140 is in master I think and perhaps you are saying there are more changes in master (which is true from the diff)?

Ah, yes. It looks some changes are intendedly not into v1.0.1 (vs the master), and given the results from the linter in my local, v1.0.1 missed some important changes. Although I am not used to the policy there, I see the diff in the commit logs (master vs tag v1.0.1).

**Comment by HyukjinKwon** [comment]
> 5 min: this is mildly concerning, is it possible this is caused by new checks in lintr? perhaps we could exclude them or something?

1.0.1 miss the linters below (vs master):

```
nonportable_path_linter: check that file.path() is used to construct safe and portable paths.
pipe_continuation_linter: Check that each step in a pipeline is on a new line, or the entire pipe fits on one line.
extraction_operator_linter: check that the [[ operator is used when extracting a single element from an object, not [ (subsetting) nor $ (interactive use).
implicit_integer_linter: check that integers are explicitly typed using the form 1L instead of 1.
object_name_linter: check that object names conform to a single naming style, e.g. snake_case or lowerCamelCase.
semicolon_terminator_linter: check that no semicolons terminate statements.
todo_comment_linter: check that the source contains no TODO comments (case-insensitive).
T_and_F_symbol_linter: avoid the symbols T and F (for TRUE and FALSE).
undesirable_function_linter: report the use of undesirable functions, e.g. options or sapply and suggest an alternative.
undesirable_operator_linter: report the use of undesirable operators, e.g. ::: or <<- and suggest an alternative.
unneeded_concatenation_linter: check that the c function is not used without arguments nor with a single constant
```

I manually disabled the linters above but the elapsed time looks almost the same. So, I think some existing linters were slowed down in the master - I tried to disable everything except for each linter - https://gist.github.com/HyukjinKwon/1f852bcdcc2a13f396dfd2d6a88bea16

**Comment by HyukjinKwon** [comment]
I also tried v1.0.1 against the master; however, it looks not detecting many instances this PR fixes. It only detected 3 extra instances:

```
R/DataFrame.R:2615:22: style: Variable and function names should not be longer than 30 characters.
            colsX <- generateAliasesForIntersectedCols(x, by, suffixes[1])
                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
R/DataFrame.R:2616:22: style: Variable and function names should not be longer than 30 characters.
            colsY <- generateAliasesForIntersectedCols(y, by, suffixes[2])
                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
R/DataFrame.R:2666:1: style: Variable and function names should not be longer than 30 characters.
generateAliasesForIntersectedCols <- function (x, intersectedColNames, suffix) {
^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
lintr checks failed.
```

 So, I would like to propose to use jimhester/lintr@5431140 if possible.

I opened an issue to ask a question about this slowdown - `https://github.com/jimhester/lintr/issues/270`. Probably, we could wait for few days.  if nothing could be done for now, I guess we could consider just use jimhester/lintr@5431140 as is ...

**Comment by HyukjinKwon** [comment]
retest this please

File: R/pkg/R/DataFrame.R
Comment (felixcheung) [inline R/pkg/R/DataFrame.R]: nit: I'd remove this `@note` too

**Review by felixcheung — APPROVED** [review]

**Comment by felixcheung** [comment]
oh, remember to remove WIP and update this line https://github.com/apache/spark/pull/19290/files#diff-74ca2b618d236bbd6faa23e13bff1403R30 before merging in?

File: dev/lint-r.R
Comment (HyukjinKwon) [inline dev/lint-r.R]: I checked this - when downgrading, upgrading and not installed.

File: R/pkg/R/column.R
Comment (HyukjinKwon) [inline R/pkg/R/column.R]: I just double checked the links.

**Comment by HyukjinKwon** [comment]
Ugh.. it failed to install due to permission issue ... 

```
Downloading GitHub repo jimhester/lintr@5431140
from URL https://api.github.com/repos/jimhester/lintr/zipball/5431140
Installing lintr
Downloading GitHub repo hadley/xml2@master
from URL https://api.github.com/repos/hadley/xml2/zipball/master
Installing xml2
'/usr/lib64/R/bin/R' --no-site-file --no-environ --no-save --no-restore  \
  --quiet CMD INSTALL  \
  '/tmp/RtmprFnaBu/devtools438c516b7717/r-lib-xml2-5799cd9'  \
  --library='/usr/lib64/R/library' --install-tests 

Error: ERROR: no permission to install to directory '/usr/lib64/R/library'
Installation failed: Command failed (1)
'/usr/lib64/R/bin/R' --no-site-file --no-environ --no-save --no-restore  \
  --quiet CMD INSTALL  \
  '/tmp/RtmprFnaBu/devtools438c286c3792/jimhester-lintr-5431140'  \
  --library='/usr/lib64/R/library' --install-tests 

Error: ERROR: no permission to install to directory '/usr/lib64/R/library'
Installation failed: Command failed (1)
```

**Comment by HyukjinKwon** [comment]
Hi @shaneknapp. I am sorry but it's me again ...

Here, this PR tries to upgrade an R package, [lintr](https://github.com/jimhester/lintr) for static Code analysis for R, which is ran via `./dev/lint-r` -> `dev/lint-r.R` as a part of Jenkins PR build.

This upgrade catches many meaningful instances that should have caught before in SparkR codes and if I understood correctly, we decided to upgrade this if it does not cause a problem.

I tried to fix the `dev/lint-r.R` script so that `lintr` can be upgraded if this package is not installed or not the specific version (1.0.1.9000) after testing this in my local, but looks failed in Jenkins due to a permission issue.

I believe the bash command below installs the package correctly:

```R
Rscript -e "devtools::install_github('jimhester/lintr@5431140')"
```

Maybe, I should have cc'ed you here first before trying this one. I appologise and I will cc you in such cases in the future.

**Comment by shaneknapp** [comment]
@HyukjinKwon -- you will absolutely not have builds install packages on the build system.  this is a really bad idea.

is this absolutely required, or just to fix a warning in the build output?

**Comment by HyukjinKwon** [comment]
@shaneknapp Sure, it was my bad. I will be careful next time.

It is required to fix an actual issue in order to to detect R codes that do not follow project's R style.

**Comment by HyukjinKwon** [comment]
Do you maybe have some worries about this? If that worry is quite crucial, I think we could also consider an option, not upgrading this, leaving `lint-r.R` script as was, and only fixing the instances detected by the upgrade here.

**Comment by HyukjinKwon** [comment]
For more context, I believe `lintr` was initially installed in https://github.com/apache/spark/commit/004f57374b98c4df32d9f1e19221f68e92639a49. Upgrade to jimhester/lintr@a769c0b was proposed here in https://github.com/apache/spark/commit/7d1175011c976756efcd4e4e4f70a8fd6f287026 but I believe this didn't actually upgrade the package on Jenkins as `lintr` was already installed  and `"lintr" %in% row.names(installed.packages())  == FALSE` gives `FALSE`.

So, I believe `lintr` in Jenkins is already not synced to jimhester/lintr@a769c0b as fixed in https://github.com/apache/spark/commit/7d1175011c976756efcd4e4e4f70a8fd6f287026 unless it was manually upgraded afterwards (I can't check but this is just my assumption).

I believe eventually we should upgrade and match with this anyway to detect project's R style in each PR. I at least have been in a PR where I had to manually suggest to fix.

**Comment by felixcheung** [comment]
hmm, perhaps open a ticket to ask shane to update Jenkins to this commit checksum manually?

**Comment by HyukjinKwon** [comment]
Let me revert "... || packageVersion("lintr") != "1.0.1.9000" this condition back so that it does not try to upgrade it.

**Comment by HyukjinKwon** [comment]
Oh! but if we upgrade it first, many builds in PRs should be broken due to this check failure. Probably, let me revert the script first to not upgrade it in order to unblock this.

**Comment by HyukjinKwon** [comment]
I just double checked the newer lintr, lintr@5431140, passes on the top of the master with the current change.

**Comment by HyukjinKwon** [comment]
retest this please

**Review by felixcheung — COMMENTED** [review]
yap, let's update the PR title too

**Comment by shaneknapp** [comment]
hey all, i'm here.  was out sick for the past few days and trying to get
caught up.  sorry about that!

so...  what version of lintr do we need to put on the workers?

**Comment by HyukjinKwon** [comment]
Ah, thank you @shaneknapp. The command below:

```
Rscript -e "devtools::install_github('jimhester/lintr@5431140')"
```

should upgrades `lintr` to the desired version here.

Actually, looks we should merge this one first before upgrading `lintr` in Jenkins. I realised upgrading it first could make R lintr check failures in other PRs up to my knowledge ..

I updated the change here back to not try to upgrade `lintr`. So, merging this one would not cause a problem on the build system. Let me go ahead with this one first and then I guess we could upgrade.

Thanks again and I am sorry for this, @shaneknapp.

**Comment by felixcheung** [comment]
btw, lintr once upgraded will run on all builds on all branches right? wouldn't the upgrade break other branches?

**Comment by shaneknapp** [comment]
@felixcheung  --  yes, this is the system default lintr, meaning all calls to lintr will be against this version.

as for other branches, i think it could possibly break them.

@shivaram thoughts?

**Comment by HyukjinKwon** [comment]
Let me merge this one first. This shouldn't cause any problem to built system for now.

**Comment by HyukjinKwon** [comment]
Merged to master.

**Comment by HyukjinKwon** [comment]
So, @felixcheung, @shaneknapp and @shivaram, looks we have comments, https://github.com/apache/spark/pull/19290#issuecomment-333321991 and https://github.com/apache/spark/pull/19290#issuecomment-333322653, to discuss further.

The problem looks upgrading `lintr` to `jimhester/lintr@5431140` could break other builds in other branches.

Should I rather try to install this for each build, for example, via checking the environment variables as a workaround? I checked those before - https://github.com/apache/spark/pull/17669#issue-222376466

**Comment by shaneknapp** [comment]
i filed a bug about this on the spark jira:
https://issues.apache.org/jira/browse/SPARK-22996

as we're about to move all the spark builds to new ubuntu machines, w/much more up2date package installs (especially R packages), we need to fix the lint problems and help make builds green again.

**Comment by felixcheung** [comment]
Shane is this going to affect one particular branch (eg. 2.3.0), or is it going to be all branches and all test runs?

The changes are fairly substantial - if we need to back port then it could be quite a bit of work

**Comment by shaneknapp** [comment]
the driving force behind this is that the version of R (and all associated
packages) is really quite old on the centos workers (3.1.1), and i'd really
like us to get as up to date as possible on the new ubuntu machines.

so far, lintr and testthat (opening a bug on that one soon) are the only
packages that haven't played nice after the upgrade.

if lintr 1.0.0.9001 is "good enough", i'd be ok w/locking the install to
this version.

**Comment by felixcheung** [comment]
Given that we are forking 2.3 and locking down the branch any time now, it might make sense to stay with the "current version running on old centos workers", even though they are old.

**Comment by HyukjinKwon** [comment]
BTW, I believe we are testing it with R 3.4.1 via AppVeyor too. I have been thinking it's good to test both old and new versions ...  I think we have a weak promise for `R 3.1+`  - http://spark.apache.org/docs/latest/index.html#downloading

**Comment by shaneknapp** [comment]
ok sounds good -- we'll keep things 'old' for now.

**Comment by shivaram** [comment]
The minimum R version supported is something that we can revisit though. I think we do this for Python, Java versions as well in the project

**Comment by felixcheung** [comment]
Right we could bump the supported R version for the next release. It should have minimal impact (since we are testing the close to the latest on appveyor... somewhat internally)

lintr is quite a bit more complicated.

Shane what’s the issue with testthat?

**Comment by HyukjinKwon** [comment]
I believe the issue with testthat he meatn is, it fails to run tests with testthat 2.0.0 is released. Just for a reminder, I rushed to update `appveyor.yml` to use a fixed version https://github.com/apache/spark/pull/20003 as it broke the tests. There are some details there to remember back.

**Comment by felixcheung** [comment]
argh, thanks for the reminder and the fix.
I knew calling internal method is going to bite us ... :(

**Pull request closed without merging**
